segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/metadata/MultiKeyReaderWriterScheduler.java

```diff
@@ -0,0 +1,196 @@
+    static class MultiKeyReaderWriterAsyncLock {
+        /**
+         * Keys to synchronize on.
+         */
+        @Getter
+        private final String[] keys;
+
+        /**
+         * Indicates whether the lock is a reader lock or a writer lock.
+         */
+        @Getter
+        private final boolean isReadonly;
+
+        /**
+         * Reference to the scheduler.
+         */
+        private final MultiKeyReaderWriterScheduler scheduler;
+
+        /**
+         * The future is completed when all keys for this lock become available.
+         */
+        private CompletableFuture<Void> readyFuture;
+
+        /**
+         * The future which is completed when lock is released.
+         */
+        @Getter
+        private final CompletableFuture<Void> doneFuture = new CompletableFuture<>();
+
+        /**
+         * Returns a CompletableFuture that will be completed when the lock is obtained.
+         * @return CompletableFuture which will be complete when the lock is obtained.
+         */
+        CompletableFuture<Void> lock() {
+            if (isReadonly) {
+                return scheduler.scheduleForRead(this);
+            } else {
+                return scheduler.scheduleForWrite(this);
+            }
+        }
+
+        /**
+         * Releases the lock.
+         */
+        void unlock() {
+            scheduler.release(this);
+        }
+    }
+
+    /**
+     * Adds a reference for the key.
+     */
+    private SchedulerData addReference(String key) {
+        synchronized (keyToDataMap) {
+            SchedulerData schedulerData = keyToDataMap.get(key);
+            // Add if this is a new key.
+            if (null == schedulerData) {
+                schedulerData = new SchedulerData();
+                keyToDataMap.put(key, schedulerData);
+            }
+            // Increment ref count.
+            schedulerData.count++;
+            return schedulerData;
+        }
+    }
+
+    /**
+     * Releases a reference for the key.
+     */
+    private void releaseReference(String key) {
+        synchronized (keyToDataMap) {
+            SchedulerData schedulerData = keyToDataMap.get(key);
+            // Decrement ref count.
+            schedulerData.count--;
+            // clean up if required.
+            if (0 == schedulerData.count) {
+                keyToDataMap.remove(key);
+            }
+        }
+    }
+
+    /**
+     * Gets a read lock over given set of keys.
+     */
+    MultiKeyReaderWriterAsyncLock getReadLock(String[] keys) {
+        return new MultiKeyReaderWriterAsyncLock(keys, true, this);
+    }
+
+    /**
+     * Gets a write lock over given set of keys.
+     */
+    MultiKeyReaderWriterAsyncLock getWriteLock(String[] keys) {
+        return new MultiKeyReaderWriterAsyncLock(keys, false, this);
+    }
+
+    /**
+     * Schedules the lock for read.
+     */
+    private synchronized CompletableFuture<Void> scheduleForRead(MultiKeyReaderWriterAsyncLock lock) {
```

Comment:
What are you synchronizing this on? Why does it need to be different than `keysToLock`?

Reply:
fixed